File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int answer = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        dfs(root);
        return this.answer;
    }

    /**
     * 
     * @param root
     * @return Maximum path which contains its root
     */
    private int dfs(TreeNode root) {
        if (root == null) {
            return -987654321;
        }
        if (root.left == null && root.right == null) {
            answer = Math.max(answer, root.val);
            return root.val;
        }
        var left = dfs(root.left);
        var right = dfs(root.right);
        answer = IntStream.of(
                answer,
                left + root.val,
                right + root.val,
                left + right + root.val,
                root.val
        ).max().getAsInt();
        
        return root.val + IntStream.of(left, right, 0).max().getAsInt();
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.answer = -987654321

        def dfs(root: TreeNode):
            if not root.left and not root.right:
                self.answer = max(self.answer, root.val)
                return root.val
            
            left = dfs(root.left) if root.left else -987654321
            right = dfs(root.right) if root.right else -987654321

            self.answer = max(self.answer, left + root.val, right + root.val, left + right + root.val, root.val)
            return root.val + max(left, right, 0)

        dfs(root)
        return self.answer